### PR TITLE
Move HalfCauchy,Binomial upstream

### PR DESCRIPTION
**Pull request by fritzo**
Addresses #1267

### CODE REVIEW

**Comment by jpchen** [comment]
ill wait till @neerajprad's PR bumping pytorch's version in travis to rerun these tests before merging

File: tests/distributions/conftest.py [hunk 1]
```diff
@@ -204,17 +204,6 @@
                  'test_data': [[0.4], [0.35]]}
             ],
             scipy_arg_fn=lambda loc, scale: ((), {"loc": np.array(loc), "scale": np.array(scale)})),
-    Fixture(pyro_dist=dist.HalfCauchy,
```
Comment (neerajprad) [inline tests/distributions/conftest.py]: I suppose we can keep this fixture as is? These tests are fast, and help in catching any errors due to the wrapping layers (e.g. `.expand_by`).

**Comment by fritzo** [comment]
@jpchen Yeah, I'll dump all distributions changes into this PR.

File: tests/distributions/conftest.py (hunk 1 above)
Comment (fritzo) [reply tests/distributions/conftest.py]: Sure, will do. I'll just remove the `loc` argument, which does not exist in the upstream version.

File: pyro/distributions/torch.py [hunk 2]
```diff
@@ -124,6 +139,15 @@ def expand(self, batch_shape):
             return type(self)(loc, scale, validate_args=validate_args)
 
 
+class HalfCauchy(torch.distributions.HalfCauchy, TorchDistributionMixin):
+    def expand(self, batch_shape):
+        try:
+            return super(HalfCauchy, self).expand(batch_shape)
+        except NotImplementedError:
+            scale = self.scale.expand(batch_shape)
+            return type(self)(scale)
```
Comment (fehiepsi) [inline pyro/distributions/torch.py]: @fritzo We also have HalfNormal :)

Comment (fritzo) [reply pyro/distributions/torch.py]: done.

**Comment by neerajprad** [comment]
This should be covered in our `pytorch-1.0` branch now.

**Pull request closed without merging**